### PR TITLE
Bogavril/extensibility3

### PR DESCRIPTION
**Pull request by bgavrilMS**
Fixes - SAL + MSAL integration

I closed off a similar PR after comments - https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/pull/3065
**Please see each commit separately (commit 4 is pretty big as it's a refactoring)**

**Testing**
ut + int

**Performance impact**
nothing added to critical path

### CODE REVIEW

File: src/client/Microsoft.Identity.Client/Extensibility/AcquireTokenForClientBuilderExtensions.cs [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+    {
+        /// <summary>
+        /// Binds the token to a key in the cache. L2 cache keys contain the key id.
+        /// No cryptographic operations is performed on the token.
+        /// </summary>
+        /// <param name="builder"></param>
+        /// <param name="keyId">A key id to which the access token is associated. The token will not be retrieved from the cache unless the same key id is presented. Can be null.</param>
+        /// <returns></returns>
+        /// <exception cref="ArgumentNullException"></exception>
+        public static AcquireTokenForClientParameterBuilder WithKeyId(
```
Comment (bgavrilMS) [inline src/client/Microsoft.Identity.Client/Extensibility/AcquireTokenForClientBuilderExtensions.cs:22]: @jmprieur - second part of extensibility

@pmaytak @trwalke - this was already reviewed as part of the other PR, but I did not find a good way to commit just this part. Apollogies for the churn.

File: src/client/Microsoft.Identity.Client/Extensibility/AbstractConfidentialClientAcquireTokenParameterBuilderExtension.cs [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+    {
+        /// <summary>
+        /// Intervenes in the request pipeline, by executing a user provided delegate before MSAL makes the token request. 
+        /// The delegate can modify the request payload by adding or removing  body parameters and headers. <see cref="OnBeforeTokenRequestData"/>
+        /// </summary>
+        /// <typeparam name="T"></typeparam>
+        /// <param name="builder">The builder to chain options to</param>
+        /// <param name="onBeforeTokenRequestHandler">An async delegate which gets invoked just before MSAL makes a token request</param>
+        /// <returns>The builder to chain other options to.</returns>
+        public static AbstractAcquireTokenParameterBuilder<T> OnBeforeTokenRequest<T>(
```
Comment (bgavrilMS) [inline src/client/Microsoft.Identity.Client/Extensibility/AbstractConfidentialClientAcquireTokenParameterBuilderExtension.cs:22]: @jmprieur - this is the main extensiblity option, as discussed. The `OnBeforeTokenRequestData` is 2 files down - it's a class made up of: 
```csharp
 /// <summary>
        /// Parameters which will be sent in the request body, as POST parameters.
        /// </summary>
        public IDictionary<string, string> BodyParameters { get; }

        /// <summary>
        /// Headers which will be sent with the request.
        /// </summary>
        public IDictionary<string, string> Headers { get; }

        /// <summary>
        /// The token endpoint, including any query parameters, where the request is being sent to.
        /// </summary>
        public Uri RequestUri { get; }
```

File: tests/Microsoft.Identity.Test.Integration.netfx/HeadlessTests/LegacyPopTest.cs [hunk 3]
```diff
@@ -0,0 +1,189 @@
...
+            RsaSecurityKey popKey = CreateRsaSecurityKey();
+
+            var cca = ConfidentialClientApplicationBuilder
+                .Create(settings.ClientId)
+                .WithAuthority(settings.Authority, true)
+                .WithExperimentalFeatures(true)
+                .WithTestLogging()
+                .Build();
+
+            var result = await cca.AcquireTokenForClient(settings.AppScopes)
```
Comment (bgavrilMS) [inline tests/Microsoft.Identity.Test.Integration.netfx/HeadlessTests/LegacyPopTest.cs:42]: @jmprieur @trwalke @pmaytak - this is an example of how the 2 extensiblity APIs can be used.

File: tests/Microsoft.Identity.Test.Integration.netfx/HeadlessTests/LegacyPopTest.cs [hunk 4]
```diff
@@ -0,0 +1,189 @@
...
+                .WithKeyId(popKey.KeyId)
+                .OnBeforeTokenRequest((data) =>
+                    {
+                        ModifyRequestWithLegacyPop(data, settings, clientCredsCert, popKey);
+                        return Task.CompletedTask;
+                    })
+                .ExecuteAsync().ConfigureAwait(false);
+
+            Assert.AreEqual(TokenSource.IdentityProvider, result.AuthenticationResultMetadata.TokenSource);
```
Comment (bgavrilMS) [inline tests/Microsoft.Identity.Test.Integration.netfx/HeadlessTests/LegacyPopTest.cs:51]: And yes, this legacy POP works on production ESTS.

File: src/client/Microsoft.Identity.Client/Extensibility/AcquireTokenForClientBuilderExtensions.cs (hunk 1 above)
Comment (jmprieur) [reply src/client/Microsoft.Identity.Client/Extensibility/AcquireTokenForClientBuilderExtensions.cs:22]: This is a public API?

Comment (bgavrilMS) [reply src/client/Microsoft.Identity.Client/Extensibility/AcquireTokenForClientBuilderExtensions.cs:22]: Yes, this is a public api, hidden in the "extensibility" namespace. I want to avoid "friend assemblies" because internal APIs are easily broken.

File: tests/Microsoft.Identity.Test.Integration.netfx/HeadlessTests/LegacyPopTest.cs [hunk 5]
```diff
@@ -0,0 +1,189 @@
...
+            at = (cca.AppTokenCache as ITokenCacheInternal).Accessor.GetAllAccessTokens().Single();
+            Assert.AreEqual(at.KeyId, popKey.KeyId);
+
+            RsaSecurityKey popKey2 = CreateRsaSecurityKey();
+
+            result = await cca.AcquireTokenForClient(settings.AppScopes)
+                .WithKeyId(popKey2.KeyId)
+                .OnBeforeTokenRequest((data) =>
+                {
+                    ModifyRequestWithLegacyPop(data, settings, clientCredsCert, popKey);
```
Comment (pmaytak) [inline tests/Microsoft.Identity.Test.Integration.netfx/HeadlessTests/LegacyPopTest.cs:73]: `popKey2`?

File: tests/Microsoft.Identity.Test.Unit/ApiConfigTests/AcquireTokenOnBehalfOfBuilderTests.cs [hunk 6]
```diff
@@ -1,98 +0,0 @@
-﻿// Copyright (c) Microsoft Corporation. All rights reserved.
```
Comment (pmaytak) [inline tests/Microsoft.Identity.Test.Unit/ApiConfigTests/AcquireTokenOnBehalfOfBuilderTests.cs:1]: Why is this removed?

File: tests/Microsoft.Identity.Test.Unit/PublicApiTests/ClientCredentialWithCertTest.cs [hunk 7]
```diff
@@ -155,15 +155,18 @@ public async Task JsonWebTokenWithX509PublicCertSendCertificateTestSendX5cCombin
...
-        [DataRow(false, null, false)]
-        [DataRow(null, null, false)] // the default is false
-        [DataRow(null, true, true)]
-        [DataRow(null, false, false)]
-        [DataRow(true, true, true)]
-        [DataRow(false, false, false)]
-        [DataRow(true, false, false)] // request overrides
-        [DataRow(false, true, true)] // request overrides
-        public async Task JsonWebTokenWithX509PublicCertSendCertificateWithClaimsTestSendX5cCombinationsAsync(bool? appFlag, bool? requestFlag, bool expectX5c)
+        //[DataRow(false, null, false)]
```
Comment (pmaytak) [inline tests/Microsoft.Identity.Test.Unit/PublicApiTests/ClientCredentialWithCertTest.cs:158]: Why are these commented?

File: src/client/Microsoft.Identity.Client/Extensibility/AbstractConfidentialClientAcquireTokenParameterBuilderExtension.cs (hunk 2 above)
[the remark this answers is not in the document]
Comment (bgavrilMS) [reply src/client/Microsoft.Identity.Client/Extensibility/AbstractConfidentialClientAcquireTokenParameterBuilderExtension.cs:22]: I was thinking about this but I think "event" like objects are more like "ON_X_HANDLER", as suggested by ASP.NET Core 

https://docs.microsoft.com/en-us/dotnet/api/microsoft.aspnetcore.authentication.openidconnect.openidconnectevents.onredirecttoidentityprovider?view=aspnetcore-6.0

File: src/client/Microsoft.Identity.Client/OAuth2/OAuth2Client.cs [hunk 8]
```diff
@@ -90,6 +107,12 @@ internal async Task<T> ExecuteRequestAsync<T>(Uri endPoint, HttpMethod method, R
             {
                 if (method == HttpMethod.Post)
                 {
+                    if (onBeforePostRequestData!=null)
+                    {
+                        var requestData = new OnBeforeTokenRequestData(_bodyParameters, _headers, endpointUri, requestContext.UserCancellationToken);
+                        await onBeforePostRequestData(requestData).ConfigureAwait(false);
+                    }
+
                     response = await _httpManager.SendPostAsync(endpointUri, _headers, _bodyParameters, requestContext.Logger)
```
[the remark this answers is not in the document]
Comment (bgavrilMS) [reply src/client/Microsoft.Identity.Client/OAuth2/OAuth2Client.cs:116]: Good catch!
